hathor/indexes/partial_rocksdb_tips_index.py

```diff
@@ -0,0 +1,155 @@
+    """ Implementation of progress helper for using with loading the interval tree.
+
+    This is basically a stripped down version of `hathor.util.progress`
+    """
+    t_start = time.time()
+    count = 0
+    count_log_prev = 0
+    log.debug('load will start')
+    t_log_prev = t_start
+    while True:
+        try:
+            iv = next(iter_iv)
+        except StopIteration:
+            break
+
+        t_log = time.time()
+        dt_log = LogDuration(t_log - t_log_prev)
+        if dt_log > _DT_LOG_PROGRESS:
+            t_log_prev = t_log
+            dcount = count - count_log_prev
+            rate = '?' if dt_log == 0 else dcount / dt_log
+            kwargs = dict(rate=rate, iv_new=dcount, dt=dt_log, total=count)
+            if total is not None:
+                progress = count / total
+                # TODO: we could add an ETA since we know the total
+                log.info(f'loading... {math.floor(progress * 100):2.0f}%', progress=progress, **kwargs)
+            else:
+                log.info('loading...', **kwargs)
+            count_log_prev = count
+        count += 1
+
+        yield iv
+
+    t_final = time.time()
+    dt_total = LogDuration(t_final - t_start)
+    rate = '?' if dt_total == 0 else count / dt_total
+    log.info('loaded', count=count, rate=rate, total_dt=dt_total)
+
+
+class PartialRocksDBTipsIndex(MemoryTipsIndex, RocksDBIndexUtils):
+    """ Partial memory-rocksdb implementation
+
+    """
+
+    # An interval tree used to know the tips at any timestamp.
+    # The intervals are in the form (begin, end), where begin is the timestamp
+    # of the transaction, and end is the smallest timestamp of the tx's children.
+    tree: IntervalTree
+
+    # It is a way to access the interval by the hash of the transaction.
+    # It is useful because the interval tree allows access only by the interval.
+    tx_last_interval: Dict[bytes, Interval]
+
+    def __init__(self, db: 'rocksdb.DB', name: str) -> None:
+        MemoryTipsIndex.__init__(self)
+        RocksDBIndexUtils.__init__(self, db, f'tips-{name}'.encode())
+        self._name = name
+
+    def get_db_name(self) -> Optional[str]:
+        return f'tips_{self._name}'
+
+    def force_clear(self) -> None:
+        super().force_clear()
+        self.clear()
+
+    def _to_key(self, interval: Interval) -> bytes:
+        import struct
+        assert len(interval.data) == 32
+        begin = _to_db_value(interval.begin)
+        end = _to_db_value(interval.end)
+        return struct.pack('>II', begin, end) + interval.data
+
+    def _from_key(self, key: bytes) -> Interval:
+        import struct
+        assert len(key) == 4 + 4 + 32
+        begin, end = struct.unpack('>II', key[:8])
+        tx_id = key[8:]
+        assert len(tx_id) == 32
+        return Interval(_from_db_value(begin), _from_db_value(end), tx_id)
+
+    def init_start(self) -> None:
```

Comment:
I think a docstring in some of these methods could facilitate reading the code for people not used to it.

Reply:
Oh, definitely! I'll add it.